plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,282 @@
+
+        log.debug(String.format(Locale.ROOT, "successfully ingested record type: %s", metricType));
+    }
+
+    private synchronized void checkAndResetWindow(MetricType metricType, Long timestamp) {
+        Long windowStart = calculateWindowStart(timestamp);
+        // reset window if the current window is outdated
+        if (!topQueriesWindowStart.containsKey(metricType) || topQueriesWindowStart.get(metricType) < windowStart) {
+            getStoreByMetricType(metricType).clear();
+            topQueriesWindowStart.put(metricType, windowStart);
+        }
+    }
+
+    /**
+     * Get all top queries records that are in the query insight store, based on the input MetricType
+     * By default, return the records in sorted order.
+     *
+     * @param metricType {@link MetricType}
+     * @return List of the records that are in the query insight store
+     * @throws IllegalArgumentException if query insight is disabled in the cluster
+     */
+    public List<SearchQueryRecord> getTopNRecords(MetricType metricType) throws IllegalArgumentException {
+        if (!enableCollect.containsKey(metricType) || !enableCollect.get(metricType)) {
+            throw new IllegalArgumentException(
+                String.format(
+                    Locale.ROOT,
+                    "Cannot get query data when query insight feature is not enabled for MetricType [%s].",
+                    metricType
+                )
+            );
+        }
+        checkAndResetWindow(metricType, System.currentTimeMillis());
+        PriorityBlockingQueue<SearchQueryRecord> store = getStoreByMetricType(metricType);
+        List<SearchQueryRecord> queries = new ArrayList<>(store);
+        queries.sort((a, b) -> SearchQueryRecord.compare(a, b, metricType) * -1);
+        return queries;
+    }
+
+    /**
+     * Set flag to enable or disable Query Insights data collection
+     *
+     * @param metricType {@link MetricType}
+     * @param enableCollect Flag to enable or disable Query Insights data collection
+     */
+    public void setEnableCollect(MetricType metricType, boolean enableCollect) {
+        this.enableCollect.put(metricType, enableCollect);
+        // set topQueriesWindowStart to enable top n queries collection
+        if (enableCollect) {
+            topQueriesWindowStart.put(metricType, calculateWindowStart(System.currentTimeMillis()));
+        }
+    }
+
+    /**
+     * Get if the Query Insights data collection is enabled for a MetricType
+     *
+     * @param metricType {@link MetricType}
+     * @return if the Query Insights data collection is enabled
+     */
+    public boolean getEnableCollect(MetricType metricType) {
+        return this.enableCollect.containsKey(metricType) && this.enableCollect.get(metricType);
+    }
+
+    /**
+     * Set the top N size for TopQueriesByLatencyService service.
+     *
+     * @param size the top N size to set
+     */
+    public void setTopNSize(int size) {
```

Comment:
should we rename it to capacity instead of topN?

Reply:
The top N queries is one of the use cases in this query insights service, although there's only one store (the priorityQueue), there could be more stores in the future. Ideally this function should only set the capacity for the PriorityQueue, renaming to capacity could cause some confusion.